### DISCUSSION TITLE
Why does Patrick copy the code of some of the Openzepplin contracts to local?

### DISCUSSION THREAD

**Discussion by JimAtActiv** [smartcontractkit/full-blockchain-solidity-course-py#1160] [Q&A] [answered]
Hello All-

While taking the course, I see that on some occasions Patrick copies the source code for some of the Openzeppelin contracts into his local project.  I am curious why he did not just simply import the contract from Openzeppelin's GitHub?

Thanks!

Jim

**Comment by n4n0b1t3** [comment, marked as answer]
Can you state the lesson and, even better, send a Youtube link with the exact timestamp?

**Reply by JimAtActiv** [reply]
In Lesson 12 at timestamp https://youtu.be/M576WGiDBdQ?t=43562 Patrick shows going to [OpenZeppelin's Github](https://github.com/OpenZeppelin/openzeppelin-contracts/tree/master/contracts/proxy/transparent) and coping the source code of two contracts, ProxyAdmin.sol and TransparentUpgradeableProxy.sol.  I am ASSUMING, and maybe this is where I am wrong, that these contracts are included in the OpenZeppelin package.  If they are, it is not clear to me why he did not just import  the contracts in question from the package in the same manor he imported the ERC721 contract from Zeppelin.  I'm sure there is a good reason, just curious about what it might be.

Thanks!

Jim

**Reply by JimAtActiv** [reply]
I just dug around a bit within my brownie project (I am using MS Code) and I see that under build/contracts/dependenciew/OpenZeppelin there are a bunch of ABI json files for eight contracts.  Does this mean that when you import the OpenZeppelin package, it does not include all of the contracts found under https://github.com/OpenZeppelin/openzeppelin-contracts.

If this is the case, is there an easy way to know which contracts are included within a package?  Is looking under the build folder the best way?  Does the build folder only list the contract ABIs for the contracts I'm actually using in my project?

**Reply by n4n0b1t3** [reply]
Just a side note: Under windows, Brownie is creating a directory at C:\Users\Username\.brownie\ Look into packages. you should find Openzeppelin.
To your original question. I noticed this in other tutorials as well and often it was related to the code being deprecated and not provided any more in the most recent releases. This might have been true at the time when this video was created, since transparent proxies have been replaced since then with UUPS. see here https://docs.openzeppelin.com/contracts/4.x/api/proxy#transparent-vs-uups

**Reply by JimAtActiv** [reply]
Great, Patrick did mention that UUPS was more often being used now, which probably explains why he copied the original source code.  Appreciate the answer!
